DNN Platform/DotNetNuke.Web/Common/RequestScopeServiceProvider.cs

```diff
@@ -0,0 +1,75 @@
+﻿namespace DotNetNuke.Web.Common
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Reflection;
+    using System.Web;
+
+    using DotNetNuke.Common;
+    using DotNetNuke.Common.Extensions;
+    using Microsoft.Extensions.DependencyInjection;
+
+    /// <summary>
+    /// Service provider implementation for <see cref="HttpRuntime.WebObjectActivator"/>.
+    /// </summary>
+    internal class RequestScopeServiceProvider : IServiceProvider
+    {
+        private const int TypesCannotResolveCacheCap = 100000;
+        private readonly ConcurrentDictionary<Type, bool> typesCannotResolve = new ConcurrentDictionary<Type, bool>();
+
+        /// <inheritdoc />
+        public object GetService(Type serviceType)
+        {
+            if (this.typesCannotResolve.ContainsKey(serviceType))
+            {
+                return DefaultCreateInstance(serviceType);
+            }
+
+            var provider = HttpContextSource.Current?.GetScope()?.ServiceProvider;
```

Comment:
Could you explain the purpose of this class as I don't understand why we need to have 2 separate implementations of the `IServiceProvider` in DNN.

This appears to be a wrapper around the `IServiceProvider` that is stored on the on the HttpContext. If you are looking for more details on how this is set, you can look at the `ServiceRequestScopeModule` which adds this onto the HttpContext. This then allows you to reference the `IServiceProvider` implementation here to resolve dependencies.

My questions is why can't we just use the `LazyServiceProvider` as suggested above?

Reply:
Every request a new scope is created by the `ServiceRequestScopeModule`. The new `RequestScopeServiceProvider` uses that scope and resolves the services requested by controls.

If we would use the instance of `LazyServiceProvider` here the services are being created in the global scope, not the request scope; causing them never to dispose. 

Additionally, whenever a service could not be found (for instance a HttpModule) it creates a new instance of that service. This is something that the `LazyServiceProvider` doesn't do, which causes ASP.Net to complain.

Reply:
This makes a lot of sense, thank you for clarifying this 👍